### PR TITLE
joelbandi-Flow tests with page objects

### LINKED ISSUE #10: Refactor flow tests to use page objects
#### As a developer I want my flow tests to be written using the object-oriented [page object pattern](http://martinfowler.com/bliki/PageObject.html).

_Assignees_: @kkajla12, @lsq921224, @ragurney

__Story__:

Now that we have gotten some experience writing flow tests with Capybara,
we've noticed that these tests a) can be brittle in that changing selectors
requires updating many tests and b) use a low-level interface that can make
reading and writing flow tests more cumbersome than we'd like.

Fortunately, there is a better way: using page objects. With page objects,
we can create objects to encapsulate pages, elements, and actions that can be
taken in our application. We can design an interface for interacting with our
application that maps directly to how our users would think about interacting
with the application. This interface makes the tests easier to read, write,
and modify, and reduces the possibility of writing flaky or invalid tests.

Working on this story will be a bit different than your typical stories. We've
created a starting point for you in the form of a branch in your repo called
`flowTestsWithPageObjects`. This branch contains flow tests for adding images,
deleting images, and viewing images associated with a tag (in the
`images_crud_test.rb` file). These tests are fully written, but the page
objects they use are not. Your job is to fill in the page objects to make the
tests pass against your application.

The first step is to rebase the `flowTestsWithPageObjects` branch on top of
your updated master branch. Note that if you already have an
`images_crud_test.rb` file, there will be a merge conflict, so you may want to
rename your file first, rebase, make the tests on that branch pass, and then
migrate any other tests that were in your file to the page object style before
deleting your old test file.

Finally, there is a video to help you get started. You should watch this video
and also take a look at the ae_page_objects readme which will be an extremely
helpful resource as you learn to write page objects. Both are linked in the
reference material below.

__Acceptance criteria__:
- [ ] The flow tests have been refactored to use page objects.

__Discussion topics__:
- [ ] What makes a good page object interface?

__Reference material__:
- https://github.com/appfolio/ae_page_objects
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects

__Dependencies__:
- Delete Images
- Images by Tag

### PR DESCRIPTION
**Pull request by joelbandi**
Closes #10 

---

__Acceptance criteria__:
- [x] The flow tests have been refactored to use page objects.

__Discussion topics__:
- [x] What makes a good page object interface?

### CODE REVIEW

File: test/page_objects/images/index_page.rb [hunk 1]
```diff
@@ -14,8 +14,8 @@ def add_new_image!
         window.change_to(NewPage)
       end
 
-      def showing_image?(url:, tags: nil)
-        #TODO
+      def showing_image?(url)
+        node.find("img[src=\"#{url}\"]").present?
```
Comment (kkajla12) [inline test/page_objects/images/index_page.rb]: This definitely works, but you're looking through the _entire_ DOM tree to find if the image you're looking for exists. Plus it's almost too concise, and therefore a little hard to understand quickly. Do you think there's a way to do this in more idiomatic Ruby? HInt: Is there a way to iterate over an existing list of items and return true if any of them matches the conditions you are expecting?

P.S. What if there are multiple images with the same URL, but each one has different tags (which you can use to differentiate between them). Let's say you're looking for Image B and it's not there. This current implementation would return true even though the actual image you're looking for isn't really there. AKA: You forgot about the tags!

File: app/views/images/index.html.erb [hunk 2]
```diff
@@ -2,7 +2,7 @@
 <%= content_tag(:h1, "Images with the tag : #{@tag_filter}", id: 'tag_filter_description') if (@tag_filter && !@images.empty?) %>
 <div class='index-image-grid'>
   <% @images.each do |image| %>
-    <%= link_to content_tag(:div, render('thumbnail', image: image), class: 'index-image-grid__thumb'), image_path(image) %>
+    <%= link_to content_tag(:div, render('thumbnail', image: image), class: 'index-image-grid__thumb'), image_path(image), class: 'index-image-grid__image-link' %>
```
Comment (kkajla12) [inline app/views/images/index.html.erb]: Nice use of BEM naming! :)

File: test/flows/images_crud_test.rb [hunk 3]
```diff
@@ -23,39 +23,39 @@ class ImagesCrudTest < FlowTestCase
     assert_equal tags, image_show_page.tags
 
     images_index_page = image_show_page.go_back_to_index!
-    assert images_index_page.showing_image?(image_url)
+    assert images_index_page.showing_image?(url: image_url)
```
Comment (kkajla12) [inline test/flows/images_crud_test.rb]: I love keyword arguments. Nice job!

File: test/page_objects/images/index_page.rb [hunk 4]
```diff
@@ -14,8 +17,10 @@ def add_new_image!
         window.change_to(NewPage)
       end
 
-      def showing_image?(url)
-        node.find("img[src=\"#{url}\"]").present?
+      def showing_image?(url:)
+        images.any? do |image|
```
Comment (kkajla12) [inline test/page_objects/images/index_page.rb]: Nice, this is much cleaner, idiomatic way to this code. You should still think about the tags. Refer to my previous comment on this.

File: test/flows/images_crud_test.rb [hunk 5]
```diff
@@ -57,27 +57,28 @@ class ImagesCrudTest < FlowTestCase
...
-  #   refute images_index_page.showing_image?(url: cat_url)
-
-  #   images_index_page = images_index_page.clear_tag_filter!
-  #   assert_equal 3, images_index_page.images.count
-  # end
+  test 'view images associated with a tag' do
+    puppy_url_1 = 'http://www.pawderosa.com/images/puppies.jpg'
+    puppy_url_2 = 'http://ghk.h-cdn.co/assets/16/09/980x490/landscape-1457107485-gettyimages-512366437.jpg'
+    cat_url = 'https://i.pinimg.com/736x/c9/dc/d4/c9dcd4334391176f134b655f01b4200b--cat-things-cat-lady.jpg'
+    Image.create!([
```
Comment (kkajla12) [inline test/flows/images_crud_test.rb]: Maybe you could save the images created here to an array and then use that array below on line 71?

**Comment by kkajla12** [comment]
This looks great overall! I left a few comments that you should definitely read through and understand. Let me know if you have any questions.

File: test/page_objects/images/index_page.rb (hunk 1 above)
Comment (joelbandi) [reply test/page_objects/images/index_page.rb]: Yeap! this was before I initialized a collection. I did refactor it in the next commit and used `any?` enum method!

Regarding the tags: The index page does not have any information about the tags of each image(I'm not displaying them) and that means in a normal index page user flow a user can never have any type of tag info about an image unless the user clicks on it. 

I did think about this and wanted to change my actual markup a little bit but I didn't want to pollute this PR

**Pull request merged**